Fail when installing from Github-generated .zip

j00bar opened the pull request:
##### SUMMARY
The .zip files generated by GitHub for download do not support symbolic links, however the `bin` directory in the source tree depends on symlinks being there to work. When installing from a GitHub generated .zip-file, when the user runs any script except `ansible` (e.g. `ansible-playbook`), they are given an error as if they were trying to run `ansible`.

This patch makes `setup.py` bomb out with a helpful error message in the case that `ansible-playbook` is not a symlink.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
setup.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0
  config file = 
  configured module search path = [u'/Users/jginsberg/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  python version = 2.7.13 (default, Dec 18 2016, 07:03:39) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]

```


##### ADDITIONAL INFORMATION
New output when installing from .zip file:
```
(github-fix) Plucky-Badger:bin jginsberg$ pip install https://github.com/j00bar/ansible/archive/fail-on-github-zipfile.zip
Collecting https://github.com/j00bar/ansible/archive/fail-on-github-zipfile.zip
  Downloading https://github.com/j00bar/ansible/archive/fail-on-github-zipfile.zip
     - 7.5MB 3.3MB/s
    Complete output from command python setup.py egg_info:
    It appears that you are installing Ansible from an archive that does not support symbolic links, usually a .zip file downloaded from GitHub. This renders your copy of Ansible unusable.
    
    
    Please re-acquire Ansible from PyPI or from the git repository, and try again.
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 3 in /private/var/folders/wg/d5z1m_tn7d31p142wgc800tw0000gp/T/pip-edBgAD-build/
```

Output when installing from git repo or PyPi remains unchanged.

j00bar committed:
Fail when installing from Github-generated .zip

kushaldas commented:
The code looks good, but when I tried to download the zip files from Github, all of them had the symlinks correctly pointing to the ansible file. I tried different branches, including 

```
https://github.com/j00bar/ansible/archive/fail-on-github-zipfile.zip
```

j00bar commented:
Interesting. This appears to be a failure in Python's zipfile implementation, not GitHub's zipfiles:

```
Big-Kahuna:wtf jginsberg$ wget https://github.com/j00bar/ansible/archive/fail-on-github-zipfile.zip
--2017-03-30 11:16:54--  https://github.com/j00bar/ansible/archive/fail-on-github-zipfile.zip
Resolving github.com... 192.30.253.113, 192.30.253.112
Connecting to github.com|192.30.253.113|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://codeload.github.com/j00bar/ansible/zip/fail-on-github-zipfile [following]
--2017-03-30 11:16:54--  https://codeload.github.com/j00bar/ansible/zip/fail-on-github-zipfile
Resolving codeload.github.com... 192.30.253.120, 192.30.253.121
Connecting to codeload.github.com|192.30.253.120|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/zip]
Saving to: ‘fail-on-github-zipfile.zip’

fail-on-github-zipf     [           <=>      ]   7.15M  3.38MB/s    in 2.1s    

2017-03-30 11:16:57 (3.38 MB/s) - ‘fail-on-github-zipfile.zip’ saved [7493654]

Big-Kahuna:wtf jginsberg$ python -m zipfile -e fail-on-github-zipfile.zip ansible/
Big-Kahuna:wtf jginsberg$ cd ansible/ansible-fail-on-github-zipfile/bin/
Big-Kahuna:bin jginsberg$ ls -l
total 96
-rw-r--r--  1 jginsberg  wheel   4883 Mar 30 11:19 ansible
-rw-r--r--  1 jginsberg  wheel  13139 Mar 30 11:19 ansible-connection
-rw-r--r--  1 jginsberg  wheel      7 Mar 30 11:19 ansible-console
-rw-r--r--  1 jginsberg  wheel      7 Mar 30 11:19 ansible-doc
-rw-r--r--  1 jginsberg  wheel      7 Mar 30 11:19 ansible-galaxy
-rw-r--r--  1 jginsberg  wheel      7 Mar 30 11:19 ansible-playbook
-rw-r--r--  1 jginsberg  wheel      7 Mar 30 11:19 ansible-pull
-rw-r--r--  1 jginsberg  wheel      7 Mar 30 11:19 ansible-vault
```

I'll update the PR with better messaging.

sivel commented:
Yes, I believe the problem is that we *don't* want them to be symlinks within a zip file.

This is a problem in how python deals with zip files containing symlinks.  If I extract a zip file via unzip, I get symlinks. If I extract a zip file using the `zipfile` python module, I get real files where the symlinks should be but with the name of the file they should symlink to, inside of the file, and not symlinks.

If you pip install the zip file from github, you get exactly this.  You end up with are files with just the string `ansible` as the contents of the file, which when executed, causes the file to just execute the `ansible` file.

If you pip install the tar.gz directly from github, things work fine.

So the problem isn't even with symlinks, or not symlinks, it is purely with dealing with a zip file containing symlinks.  The solution would need to only restrict this usage, and I am not sure it can.

j00bar committed:
Fix error message to be more accurate.

The bug lies in Python’s zipfile implementation, not in GitHub’s .zip generation.

j00bar commented:
@sivel I would speculate the absurdly overwhelming majority of people install Ansible from PyPI or using git. The impression @bcoca gave me was that this comes up only seldom. It happened to burn an hour of my time yesterday, so I figured I'd try to save other people the frustration.

Failing out on this type of install versus changing the way that the `bin/` scripts are constructed and identified seems like the appropriate balance of intrusiveness and inconvenience for how infrequently this install vector is used.

sivel commented:
I'm not sure what point you are trying to convey.  I'm not saying you shouldn't try to solve it, I am saying that so far, the way it has been solved is incorrect, and or potentially catches other non-failure scenarios.

I am merely saying that the exact scenario that causes this is a zip file with symlinks.  So you cannot simply just check for symlinks, or lack of symlinks, as there are scenarios where those things work.

You would need to match the exact scenario of zip file plus symlinks to fail on.  I'm not sure how feasible that is.  The problem is that when python extracts the zip file, they are no longer symlinks, they are real files.

I suppose you could check that the `ansible-playbook` file is a real file, and that it's contents exactly equal `ansible`, and then fail in that scenario.

j00bar commented:
Right... So, when you, say, `pip install https://github.com/ansible/ansible/archive/devel.zip` or in another app's `setup.py` list an `install_requires` including `ansible==2.4.0` and a `dependency_links` including `'https://github.com/ansible/ansible/archive/devel.zip#egg=ansible-2.4.0`, what happens is that Python pulls down that archive and decompresses it using its own `zipfile` library into build-space. Given that it's doing that decompression with `zipfile`, it loses the fact that `bin/ansible-playbook` et al are symlinks to `bin/ansible` and instead writes them as plain files whose contents are simply the text `ansible`. The next step after decompression is running `setup.py` from the archive just decompressed.

This patch looks from the `setup.py`'s execution as to whether-or-not `bin/ansible-playbook` is a symlink or not. If it's not, it errors out. If Ansible's `setup.py` is running, the .zip file has been unzipped. It's precisely what you said: "match the exact scenario of a zip file plus symlinks to fail on."

You're correct another way to do this from `setup.py` would be, instead of `not os.path.islink('bin/ansible-playbook')`, a test for `not os.path.islink('bin/ansible-playbook') and open('bin/ansible-playbook').read() == 'ansible'`.

sivel commented:
Perform the following in your branch as it is:

```
python setup.py sdist
cd dist
pip install ansible-2.4.0.tar.gz
```

This will cause the following to then happen:

```
Processing ./ansible-2.4.0.tar.gz
    Complete output from command python setup.py egg_info:
    It appears that you are installing Ansible from a .zip file. Because Python's zipfile library does not properly handle symbolic links in a .zip file, pip and setuptools cannot install Ansible like this.


    Please either unzip the .zip file yourself and try again with the uncompressed version, or install Ansible with pip or setuptools using the versions on PyPI or directly from the git repository.


    ----------------------------------------
Command "python setup.py egg_info" failed with error code 3 in /private/tmp/pip-3Keby9-build/
```

Your current solution would cause the tarball uploaded to pypi to then become un-installable.  This is because `python setup.py sdist` turns the symlink into real files.

j00bar commented:
Ah, thank you. That clarifies what you're trying to communicate perfectly: Python's sdist also does not preserve symlinks but instead resolves the symlinks into the real files they point to.

I'll amend the patch accordingly.

j00bar committed:
Fix to accomodate setup.py sdist

sdist resolves symlinked paths to what they point to.

j00bar commented:
The latest commit:

* works with `sdist` created archives
* works with `git+https://` urls
* fails with `.zip` urls

Thanks for your help!

sivel commented:
This has been resolved in a different way in https://github.com/ansible/ansible/pull/23233
